src/rules/noDuplicateVariableRule.ts

```diff
@@ -38,73 +39,56 @@ export class Rule extends Lint.Rules.AbstractRule {
     };
     /* tslint:enable:object-literal-sort-keys */
 
-    public static FAILURE_STRING_FACTORY = (name: string) => {
+    public static FAILURE_STRING(name: string): string {
         return `Duplicate variable: '${name}'`;
     }
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new NoDuplicateVariableWalker(sourceFile, this.getOptions()));
+        return this.applyWithFunction(sourceFile, walk);
     }
 }
 
-class NoDuplicateVariableWalker extends Lint.BlockScopeAwareRuleWalker<{}, ScopeInfo> {
-    public createScope(): any {
-        return null;
-    }
-
-    public createBlockScope(): ScopeInfo {
-        return new ScopeInfo();
-    }
-
-    public visitBindingElement(node: ts.BindingElement) {
-        const isSingleVariable = node.name.kind === ts.SyntaxKind.Identifier;
-        const isBlockScoped = Lint.isBlockScopedBindingElement(node);
-
-        // duplicate-variable errors for block-scoped vars are caught by tsc
-        if (isSingleVariable && !isBlockScoped) {
-            this.handleSingleVariableIdentifier(node.name as ts.Identifier);
+function walk(ctx: Lint.WalkContext<void>): void {
+    const scopes: Array<Set<string>> = [new Set()];
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isFunctionScopeBoundary(node)) {
+            scopes.push(new Set());
```

Comment:
you don't need a stack. just replace `scopes` with `scope`. Save it in a variable here, assign the new Scope and restore it 2 lines below.